### PR TITLE
ovl obj raillift OK and Mostly Documented

### PR DESCRIPTION
**Pull request by hensldm**
Before opening this PR, ensure the following:
- `./format.sh` was run to apply standard formatting.
- `make` successfully builds a matching ROM.
- No new compiler warnings were introduced during the build process.
    - Can be verified locally by running `tools/warnings_count/check_new_warnings.sh`
- New variables & functions should follow standard naming conventions.
- Comments and variables have correct spelling.
---

I was unable to find any in game that actually use `OBJRAILLIFT_SHOULD_TELEPORT` or `OBJRAILLIFT_SHOULD_REACT_TO_WEIGHT` though I might just be forgetting ones. Also happy to take any better namings.

### CODE REVIEW

File: src/overlays/actors/ovl_Obj_Raillift/z_obj_raillift.c [hunk 1]
```diff
@@ -1,3 +1,9 @@
+/*
+ * File: z_obj_raillift.c
+ * Overlay: Obj_Raillift
+ * Description: OOT Water Temple Waterfall Platform and Moving Deku Flower Platform
```
Comment (engineer124) [inline src/overlays/actors/ovl_Obj_Raillift/z_obj_raillift.c:4]:
```suggestion
 * Description: Moving Deku Flower Platform and OOT Water Temple Waterfall Platform
```
I think making the MM usage of this actor the main point, then OoT's usage as the secondary comment would be better

File: src/overlays/actors/ovl_Obj_Raillift/z_obj_raillift.c [hunk 2]
```diff
@@ -21,30 +43,242 @@ const ActorInit Obj_Raillift_InitVars = {
...
+        Math_Vec3f_Scale(&this->dyna.actor.velocity, this->dyna.actor.speedXZ / speed);
+        this->dyna.actor.world.pos.x += this->dyna.actor.velocity.x;
+        this->dyna.actor.world.pos.y += this->dyna.actor.velocity.y;
+        this->dyna.actor.world.pos.z += this->dyna.actor.velocity.z;
+    } else {
+        this->curPoint += this->direction;
+        if (1) {}
+        this->dyna.actor.speedXZ *= 0.4f;
+        shouldTeleport = OBJRAILLIFT_SHOULD_TELEPORT(&this->dyna.actor);
+        updatePos = true;
```
Comment (engineer124) [inline src/overlays/actors/ovl_Obj_Raillift/z_obj_raillift.c:161]:
```suggestion
        isPosUpdated = true;
```
Perhaps a name like this to include `is` to indicate that it's a boolean?

File: src/overlays/actors/ovl_Obj_Raillift/z_obj_raillift.c [hunk 3]
```diff
@@ -21,30 +43,242 @@ const ActorInit Obj_Raillift_InitVars = {
...
+        if (((this->curPoint >= this->endPoint) && (this->direction > 0)) || ((this->curPoint <= 0) && (this->direction < 0))) {
+            if (!shouldTeleport) {
```
Comment (engineer124) [inline src/overlays/actors/ovl_Obj_Raillift/z_obj_raillift.c:163]:
```suggestion
            if (!isTeleporting) {
```
Same concept with the bool here

File: src/overlays/actors/ovl_Obj_Raillift/z_obj_raillift.c [hunk 4]
```diff
@@ -21,30 +43,242 @@ const ActorInit Obj_Raillift_InitVars = {
...
+                this->direction = -this->direction;
+                this->waitTimer = 10;
+                this->actionFunc = ObjRaillift_Wait;
+            } else {
+                endPoint = &this->points[this->endPoint];
+                this->curPoint = (this->direction > 0) ? (0) : (this->endPoint);
```
Comment (engineer124) [inline src/overlays/actors/ovl_Obj_Raillift/z_obj_raillift.c:169]:
```suggestion
                this->curPoint = (this->direction > 0) ? 0 : this->endPoint;
```
Personal preference, I don't think you need these brackets

File: src/overlays/actors/ovl_Obj_Raillift/z_obj_raillift.c [hunk 5]
```diff
@@ -21,30 +43,242 @@ const ActorInit Obj_Raillift_InitVars = {
...
+                initialPoint = &this->points[0];
+                if (((initialPoint->x != endPoint->x) || (initialPoint->y != endPoint->y)) || (initialPoint->z != endPoint->z)) {
```
Comment (engineer124) [inline src/overlays/actors/ovl_Obj_Raillift/z_obj_raillift.c:171]:
```suggestion
                if ((initialPoint->x != endPoint->x) || (initialPoint->y != endPoint->y) || (initialPoint->z != endPoint->z)) {
```
Assuming it still matches, this should be fine

File: src/overlays/actors/ovl_Obj_Raillift/z_obj_raillift.c [hunk 6]
```diff
@@ -21,30 +43,242 @@ const ActorInit Obj_Raillift_InitVars = {
...
-#pragma GLOBAL_ASM("./asm/non_matchings/overlays/ovl_Obj_Raillift_0x80A19910/ObjRaillift_Destroy.asm")
+        if (updatePos) {
+            ObjRaillift_UpdatePosition(this, this->curPoint);
+        }
+    }
+}
 
-#pragma GLOBAL_ASM("./asm/non_matchings/overlays/ovl_Obj_Raillift_0x80A19910/func_80A19B98.asm")
+void ObjRaillift_Teleport(ObjRaillift* this, GlobalContext* globalCtx) {
+    // will teleport to what ever curpoint is set to
```
Comment (engineer124) [inline src/overlays/actors/ovl_Obj_Raillift/z_obj_raillift.c:186]:
```suggestion
/*
 * Will teleport to what ever curpoint is set to
 */
void ObjRaillift_Teleport(ObjRaillift* this, GlobalContext* globalCtx) {
```
If a comment is describing a function, then I think having like this is more standard

File: src/overlays/actors/ovl_Obj_Raillift/z_obj_raillift.c [hunk 7]
```diff
@@ -21,30 +43,242 @@ const ActorInit Obj_Raillift_InitVars = {
...
+    this->actionFunc(this, globalCtx);
+    Actor_SetHeight(&this->dyna.actor, 10.0f);
+    if (this->cutsceneTimer > 0) {
+        this->cutsceneTimer--;
+        if (this->cutsceneTimer == 0) {
+            ActorCutscene_Stop(this->dyna.actor.cutscene);
+        }
+    }
+    if (OBJRAILLIFT_SHOULD_REACT_TO_WEIGHT(thisx)) {
+        this->hadWeightOn = this->hasWeightOn;
```
Comment (engineer124) [inline src/overlays/actors/ovl_Obj_Raillift/z_obj_raillift.c:234]:
```suggestion
        this->isWeightOnPrev = this->isWeightOn;
```
Maybe something like this to keep with boolean convention?

File: src/overlays/actors/ovl_Obj_Raillift/z_obj_raillift.c [hunk 8]
```diff
@@ -21,30 +43,242 @@ const ActorInit Obj_Raillift_InitVars = {
...
+            this->unk184 = -0x8000;
+            this->unk180 = 6.0f;
+        }
+        this->unk184 += 0xCE4;
+        Math_StepToF(&this->unk180, 0.0f, 0.12f);
+        step = this->hasWeightOn ? Math_CosS(fabsf(this->unk17C) * 2048.0f) + 0.02f : Math_SinS(fabsf(this->unk17C) * 2048.0f) + 0.02f;
+        target = this->hasWeightOn ? -8.0f : 0.0f;
+        Math_StepToF(&this->unk17C, target, step);
+        this->dyna.actor.shape.yOffset = ((Math_SinS(this->unk184) * this->unk180) + this->unk17C) * 10.0f;
+        dummy_label_653499:;
```
Comment (engineer124) [inline src/overlays/actors/ovl_Obj_Raillift/z_obj_raillift.c:250]:
```suggestion
        dummy:;
```
I don't think there's a convention for this, so it's up to you. But I don't think there's a need for all the random characters that mips2c spits out once it's matching

File: src/overlays/actors/ovl_Obj_Raillift/z_obj_raillift.c [hunk 9]
```diff
@@ -21,30 +43,242 @@ const ActorInit Obj_Raillift_InitVars = {
...
+}
 
-#pragma GLOBAL_ASM("./asm/non_matchings/overlays/ovl_Obj_Raillift_0x80A19910/func_80A1A330.asm")
+void ObjRaillift_DrawDekuFlowerPlatform(Actor* thisx, GlobalContext* globalCtx) {
+    func_800BDFC0(globalCtx, D_06000208);
+}
 
-#pragma GLOBAL_ASM("./asm/non_matchings/overlays/ovl_Obj_Raillift_0x80A19910/func_80A1A360.asm")
+void ObjRaillift_DrawDekuFlowerPlatformColorful(Actor* thisx, GlobalContext* globalCtx) {
+    // The colorful platforms are the ones found in Deku Palace
```
Comment (engineer124) [inline src/overlays/actors/ovl_Obj_Raillift/z_obj_raillift.c:282]: You can probably move this comment outside the function to the style I mentioned previously. Also out of curiosity, where are the non colorful platforms? Is it only deku playground or are there more places? Any chance you could add a comment to the function above with where the non-colourful platforms are? It's okay if you don't know

File: src/overlays/actors/ovl_Obj_Raillift/z_obj_raillift.c [hunk 10]
```diff
@@ -21,30 +43,242 @@ const ActorInit Obj_Raillift_InitVars = {
...
+    }
+}
+
+void ObjRaillift_Destroy(Actor* thisx, GlobalContext* globalCtx) {
+    ObjRaillift* this = THIS;
+
+    BgCheck_RemoveActorMesh(globalCtx, &globalCtx->colCtx.dyna, this->dyna.bgId);
+}
+
+void ObjRaillift_DoNothing(ObjRaillift* this, GlobalContext* globalCtx) {
```
Comment (engineer124) [inline src/overlays/actors/ovl_Obj_Raillift/z_obj_raillift.c:114]:
```suggestion
void ObjRaillift_Noop(ObjRaillift* this, GlobalContext* globalCtx) {
```
I think `Noop` is convention for if a function doesn't do anything

**Review by engineer124 — COMMENTED** [review]
Nice work! Mainly styling things

File: src/overlays/actors/ovl_Obj_Raillift/z_obj_raillift.c (hunk 10 above)
Comment (hensldm) [reply src/overlays/actors/ovl_Obj_Raillift/z_obj_raillift.c:114]: I've seen `DoNothing` used when it's an action function that doesn't do anything.

Comment (engineer124) [reply src/overlays/actors/ovl_Obj_Raillift/z_obj_raillift.c:114]: Either should be fine then :)

**Review by engineer124 — APPROVED** [review]

File: src/overlays/actors/ovl_Obj_Raillift/z_obj_raillift.h [hunk 11]
```diff
@@ -5,9 +5,37 @@
...
+    /* 0x000 */ DynaPolyActor dyna;
+    /* 0x15C */ ObjRailliftActionFunc actionFunc;
+    /* 0x160 */ f32 speed;
+    /* 0x164 */ s32 endPoint;
+    /* 0x168 */ s32 curPoint;
+    /* 0x16C */ s32 direction; // +1 for forward, -1 for backward
+    /* 0x170 */ Vec3s* points;
+    /* 0x174 */ s32 isWeightOn;
+    /* 0x178 */ s32 isWeightOnPrev;
+    /* 0x17C */ f32 unk17C;
```
Comment (Kenix3) [inline src/overlays/actors/ovl_Obj_Raillift/z_obj_raillift.h:34]:
```suggestion
    /* 0x17C */ f32 speed;
```

This is the speed with which the lilly pad rises and falls based on Link being on or off the platform.

File: src/overlays/actors/ovl_Obj_Raillift/z_obj_raillift.h [hunk 12]
```diff
@@ -5,9 +5,37 @@
...
+    /* 0x180 */ f32 unk180;
```
Comment (Kenix3) [inline src/overlays/actors/ovl_Obj_Raillift/z_obj_raillift.h:35]:
```suggestion
    /* 0x180 */ f32 maxHeight;
```

This is the height the lilly pad is at when at full height.

File: src/overlays/actors/ovl_Obj_Raillift/z_obj_raillift.h [hunk 13]
```diff
@@ -5,9 +5,37 @@
...
+    /* 0x184 */ s16 unk184;
```
Comment (Kenix3) [inline src/overlays/actors/ovl_Obj_Raillift/z_obj_raillift.h:36]:
```suggestion
    /* 0x184 */ s16 cycle;
```
This is the sine period of the lilly pad that's reset when Link jumps on or off the lillypad.

**Review by Kenix3 — CHANGES_REQUESTED** [review]

**Review by Kenix3 — APPROVED** [review]

**Pull request merged**